Don't hash shared libraries for cache key

ChrisBr opened the pull request:
Shared libraries often contain timestamps of when they were compiled and other non-stable data
hence would often produce different hashes on different machines. This happens often in our CI.

Related to https://github.com/rubocop/rubocop/pull/8633

cc @casperisfine

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [ ] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [ ] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

casperisfine commented on lib/rubocop/result_cache.rb:
```diff
@@ -188,14 +194,24 @@ def rubocop_checksum
+      if path.end_with?(*DL_EXTENSIONS)
+        # shared libraries often contain timestamps of when
+        # they were compiled and other non-stable data
+        path
```
Hum, actually when we hash the `.rb` files, we don't care where they are located. So using the path as a "digest" mean that the `BUNDLE_PATH` now has influence, that may also be a problem for some cases. Maybe we should just extract a small name, e.g. `Zlib.crc32(File.basename(path)).to_s`.

ChrisBr replied:
Yeah good point.

jonas054 commented on lib/rubocop/result_cache.rb:
```diff
@@ -188,14 +194,24 @@ def rubocop_checksum
 
+    def digest(path)
+      if path.end_with?(*DL_EXTENSIONS)
+        # shared libraries often contain timestamps of when
```
I'd like this to be a sentence, so `Shared` instead of `shared` and a dot at the end.

jonas054 commented on CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@
 
 ## master (unreleased)
 
+* [#10841](https://github.com/rubocop/rubocop/pull/10841): Optimize cache. ([@ChrisBr][])
```
There should be a level 3 heading above this line. I guess it should be `### Changes`.

ydah replied:
The following PR template seems to be a good way to write a CHANGELOG.

> Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named {change_type}_{change_description}.md if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

jonas054 commented on lib/rubocop/result_cache.rb:
```diff
@@ -188,14 +194,24 @@ def rubocop_checksum
+        Zlib.crc32(File.basename(path)).to_s
+      else
+        content = File.binread(path)
+        Zlib.crc32(content).to_s # mtime not reliable
```
There's a little bit of duplication here. You could move `content =` to before the `if` and then have `Zlib.crc32(content).to_s` as the last line of the method, common for both branches.

jonas054 reviewed (changes requested):
Looks good to me, regarding the changed functionality. I had some minor comments.

ChrisBr commented:
♻️ @jonas054 please have another look

bbatsov commented:
Hmm, lots of tests are currently failing.

jonas054 commented:
I'm happy with the changes, but I see some builds are failing.

ydah commented:
Looks like there are some RuboCop offense in lib/rubocop/result_cache.rb 
```
Run bundle exec rake internal_investigation
RuboCop failed!
Running RuboCop...
Inspecting 1403 files
........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................W..................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

Offenses:

lib/rubocop/result_cache.rb:208:9: C: [Correctable] Layout/IndentationWidth: Use 2 (not -8) spaces for indentation.
        File.basename(path)
        ^^^^^^^^
lib/rubocop/result_cache.rb:209:7: C: [Correctable] Layout/ElseAlignment: Align else with if.
      else
      ^^^^
lib/rubocop/result_cache.rb:211:7: W: [Correctable] Layout/EndAlignment: end at 211, 6 is not aligned with if at 205, 16.
      end
      ^^^

1[40](https://github.com/rubocop/rubocop/runs/7642059901?check_suite_focus=true#step:8:41)3 files inspected, 3 offenses detected, 3 offenses autocorrectable
Error: Process completed with exit code 1.
```

ChrisBr commented:
> Hmm, lots of tests are currently failing.

Fixed my Rubocop violation and rebased on latest master but seems there are still some broken in `CI / RSpec 4 (pull_request)`. I don't think it's related to the changes in this PR though (if so I would appreciate some pointers how I can reproduce this locally, I did run `bundle exec rake default` which passed).

ydah commented on CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@
+### Changes
+
+* [#10841](https://github.com/rubocop/rubocop/pull/10841): Don't hash shared libraries for cache key. ([@ChrisBr][])
+
```
This is auto-generated by the release task and must be removed.

ChrisBr replied:
Got it, removed @ydah

ydah commented on CHANGELOG.md:
```diff
@@ -6372,3 +6376,4 @@
 [@FnControlOption]: https://github.com/FnControlOption
 [@ccutrer]: https://github.com/ccutrer
 [@Korri]: https://github.com/Korri
+[@ChrisBr]: https://github.com/ChrisBr
```
This line is also not needed as it is auto-generated by the release task. This should be removed.

koic commented:
Thanks!
